libs/isograph-react/src/core/garbageCollection.ts

```diff
@@ -51,21 +52,40 @@ export function garbageCollectEnvironment(environment: IsographEnvironment) {
     recordReachableIds(environment.store, query, retainedIds);
   }
 
-  for (const dataId in environment.store) {
-    if (!retainedIds.has(dataId)) {
-      delete environment.store[dataId];
+  for (const typeName in environment.store) {
+    const dataById = environment.store[typeName];
+    if (dataById == null) continue;
+
+    // delete all objects
+    if (!retainedIds[typeName]) {
+      delete environment.store[typeName];
+      continue;
+    }
+
+    for (const dataId in dataById) {
+      if (!retainedIds[typeName].has(dataId)) {
+        delete dataById[dataId];
+      }
+    }
+
+    if (Object.keys(dataById).length === 0) {
+      delete environment.store[typeName];
     }
   }
 }
 
+interface RetainedIds {
+  [typeName: TypeName]: Set<DataId>;
+}
+
 function recordReachableIds(
   store: IsographStore,
   retainedQuery: RetainedQuery,
-  mutableRetainedIds: Set<DataId>,
+  mutableRetainedIds: RetainedIds,
 ) {
   recordReachableIdsFromRecord(
     store,
-    store[ROOT_ID],
+    store['Query'][ROOT_ID],
```

Comment:
presumably, this would break if a user had a `schema { query: SomeotherType }` block in their schema, which is a silly thing to do, but allowed and we should handle it

Reply:
I've solved this by generating `concreteType: "Query" | "Mutation"` for each entrypoint and refetch query.
Now we have separate cache for `Query` and `Mutation`, this is better then a single `__ROOT` right?

Reply:
That's really good idea, let me think about it some more and get back to you